dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/alias/InclusionBasedPointerAnalysis.java

```diff
@@ -435,56 +471,119 @@ private Variable newVariable(String name) {
+    private List<IncludeEdge> detectCycles(Variable variable, IncludeEdge edge) {
+        // Fast check for cycles of length 1.
+        if (edge.source == variable) {
+            return List.of(edge);
+        }
+        // Fast check with lazy cycle detection:
+        // Eventually, any cycle will have a 'new' edge, where the pointer sets are equal.
+        // Therefore, we wait for this, instead of trying to immediately detect the cycle.
+        if (!equalsPointerSet(variable, edge.source)) {
+            cyclesFastCulled++;
+            return List.of();
+        }
+        // Slow check
+        Set<Variable> includerSet = new HashSet<>(List.of(variable));
+        {
+            List<Variable> queue = new ArrayList<>(List.of(variable));
+            while (!queue.isEmpty()) {
+                List<Variable> next = new ArrayList<>();
+                for (Variable current : queue) {
+                    for (Variable v : current.seeAlso) {
+                        // Culling
+                        if (includerSet.contains(v)) {
+                            continue;
+                        }
+                        // Try to find some include edge, as 'seeAlso' also indicates store and load edges.
+                        for (IncludeEdge i : v.includes) {
+                            if (i.source == current && includerSet.add(v)) {
+                                next.add(v);
+                                break;
+                            }
+                        }
+                    }
+                }
+                queue = next;
+            }
+        }
+        if (!includerSet.contains(edge.source)) {
+            cyclesSlowCulled++;
+            return List.of();
+        }
+        cyclesDetected++;
+        // Collect all cyclic paths.
+        Map<Variable, List<IncludeEdge>> edges = new HashMap<>();
+        {
+            // Use 'set' for performance.
+            Set<IncludeEdge> set = new HashSet<>();
+            List<IncludeEdge> queue = new ArrayList<>(List.of(new IncludeEdge(edge.source, TRIVIAL_MODIFIER)));
+            // Since cycles are detected lazily, we need a bound for cycle lengths.
+            for (int length = 0; !queue.isEmpty() && length < includerSet.size(); length++) {
+                List<IncludeEdge> next = new ArrayList<>();
+                for (IncludeEdge current : queue) {
+                    for (IncludeEdge i : current.source.includes) {
+                        if (edge.source != i.source && includerSet.contains(i.source)) {
+                            IncludeEdge joinedEdge = compose(i, current.modifier);
+                            if (set.add(joinedEdge) && addInto(edges.computeIfAbsent(i.source, k -> new ArrayList<>()), joinedEdge)) {
+                                next.add(joinedEdge);
+                            }
+                        }
                     }
                 }
+                queue = next;
+            }
+        }
```

Comment:
I don't know how big the time spent here is, but I believe a bidirectional BFS could be an improvement (assuming you could efficiently get outgoing includes edges, which you currently cannot).
This is just a side-note for future reference: no need to implement it.